lib/link_checker.rb

```diff
@@ -0,0 +1,73 @@
+class LinkChecker
+  class << self
+    def extract_urls(data)
+      new(data).extract_urls
+    end
+
+    def link_checker_api_bearer_token
+      ENV["LINK_CHECKER_API_BEARER_TOKEN"]
+    end
+
+    def for_files_at(path)
+      path = File.join(path, "**/*.erb") unless path.to_s.include?("*")
+      data = Dir.glob(path).map { |filename| File.read(filename) }
+      new(data)
+    end
+  end
+
+  attr_reader :data
+
+  def initialize(data)
+    @data = data
+  end
+
+  def urls
+    # Gather urls in a Set to allow simple combination and duplicate removal
+    @urls ||= extract_urls.to_set
+  end
+
+  def extract_urls
+    case data
+    when String
+      uri = URI.extract(data, %w[http https])
+
+      # markdown urls trailing bracket included in uri so removing here
+      uri.map { |u| unmatched_closing_bracket?(u) ? u.gsub(/\)\W?$/, "") : u }
+    when Array
+      urls = data.map do |datum|
+        self.class.extract_urls(datum)
```

Comment:
Slight confused as to why we need to create multiple new LinkChecker objects to rather than a single object to handle all the links?

Reply:
Do we use the `self.extract_urls` publicly? Not sure we need that class method - we could simplify the interface to this class to just `for_files_at(path)` and only deal with "data" as an array.

Reply:
I believe it's better OOP practice to have one object instant per computation rather than to reuse an instance method on a single instance. It avoid mutability issues and is a better pattern.